File: src/main/java/jp/aoichaan0513/A_TosoGame_Live/API/Manager/Player/PlayerConfig.java
package jp.aoichaan0513.A_TosoGame_Live.API.Manager.Player;

import jp.aoichaan0513.A_TosoGame_Live.API.Enums.Advancement;
import jp.aoichaan0513.A_TosoGame_Live.API.Interfaces.IConfig;
import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI;
import jp.aoichaan0513.A_TosoGame_Live.Main;
import org.bukkit.Bukkit;
import org.bukkit.configuration.file.YamlConfiguration;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;

public class PlayerConfig {

    private static UUID uuid;

    private static File file;
    private static YamlConfiguration yamlConfiguration;

    private static long money = 0;
    private static boolean hasPermission = false;
    private static boolean isBroadCaster = false;

    private static AdvancementConfig advancementConfig;

    public PlayerConfig(UUID uuid) {
        String baseFileName = "player.yml";
        String fileName = uuid.toString() + ".yml";

        PlayerConfig.uuid = uuid;

        file = new File(Main.getInstance().getDataFolder() + Main.FILE_SEPARATOR + "players" + Main.FILE_SEPARATOR + fileName);
        if (file.exists()) {
            yamlConfiguration = YamlConfiguration.loadConfiguration(file);
        } else {
            try {
                Files.copy(Main.getInstance().getResource(baseFileName), Paths.get(Main.getInstance().getDataFolder() + Main.FILE_SEPARATOR + "players" + Main.FILE_SEPARATOR + fileName));
            } catch (IOException e) {
                e.printStackTrace();
            }
            yamlConfiguration = YamlConfiguration.loadConfiguration(file);
        }

        money = yamlConfiguration.getLong("money", 0);
        hasPermission = yamlConfiguration.getBoolean("hasPermission", false);
        isBroadCaster = yamlConfiguration.getBoolean("isBroadCaster", false);

        advancementConfig = new AdvancementConfig(file, yamlConfiguration);
    }

    public UUID getUUID() {
        return uuid;
    }

    public File getFile() {
        return file;
    }

    public YamlConfiguration getConfig() {
        return yamlConfiguration;
    }


    /**
     * 所持金を取得します。
     *
     * @return 所持金
     */
    public long getMoney() {
        return money;
    }

    /**
     * 所持金を設定します。
     *
     * @param money 所持金
     */
    public void setMoney(long money) {
        PlayerConfig.money = money;
        yamlConfiguration.set("money", money);
        save();
    }

    /**
     * コマンドを使用して運営チームへの参加・離脱等が出来るかを取得します。
     *
     * @return 権限が付与されているか
     */
    public boolean getPermission() {
        return hasPermission;
    }

    /**
     * コマンドを使用して運営チームへの参加・離脱等が出来るかを設定します。
     *
     * @param hasPermission 権限が付与されているか
     */
    public void setPermission(boolean hasPermission) {
        PlayerConfig.hasPermission = hasPermission;
        yamlConfiguration.set("hasPermission", hasPermission);
        save();
    }


    /**
     * 配信者モードかどうかを取得します。
     *
     * @return 配信者モードか
     */
    public boolean getBroadCaster() {
        return isBroadCaster;
    }

    /**
     * 配信者モードかどうかを設定します。
     *
     * @param isBroadCaster 配信者モードか
     */
    public void setBroadCaster(boolean isBroadCaster) {
        PlayerConfig.isBroadCaster = isBroadCaster;
        yamlConfiguration.set("isBroadCaster", isBroadCaster);
        save();
    }

    public AdvancementConfig getAdvancementConfig() {
        return advancementConfig;
    }

    private static void setAdvancementConfig(AdvancementConfig advancementConfig) {
        PlayerConfig.advancementConfig = advancementConfig;
    }

    public static class AdvancementConfig extends IConfig {

        private final String PATH = "advancements";

        public AdvancementConfig(File file, YamlConfiguration yamlConfiguration) {
            super(file, yamlConfiguration);
        }

        public void addAdvancement(Advancement advancement) {
            if (advancement == Advancement.UNKNOWN) return;

            Set<Integer> set = new HashSet<>(yamlConfiguration.getIntegerList(PATH));
            set.add(advancement.getId());

            yamlConfiguration.set(PATH, new ArrayList<>(set));
            save();
        }

        public void removeAdvancement(Advancement advancement) {
            if (advancement == Advancement.UNKNOWN) return;

            Set<Integer> set = new HashSet<>(yamlConfiguration.getIntegerList(PATH));
            set.remove(advancement.getId());

            yamlConfiguration.set(PATH, new ArrayList<>(set));
            save();
        }

        public boolean hasAdvancement(Advancement advancement) {
            return getAdvancements().stream().anyMatch(a -> a.getId() == advancement.getId());
        }

        public Set<Advancement> getAdvancements() {
            Set<Advancement> set = new HashSet<>();
            for (int i : yamlConfiguration.getIntegerList(PATH))
                set.add(Advancement.getAdvancement(i));

            return set;
        }

        public void setAdvancements(Collection<Advancement> advancements) {
            Set<Integer> set = new HashSet<>();
            for (Advancement advancement : advancements)
                if (advancement != Advancement.UNKNOWN)
                    set.add(advancement.getId());

            yamlConfiguration.set(PATH, new ArrayList<>(set));
            save();
        }

        public void setAdvancements(Advancement... advancements) {
            Set<Integer> set = new HashSet<>();
            for (Advancement advancement : advancements)
                if (advancement != Advancement.UNKNOWN)
                    set.add(advancement.getId());

            yamlConfiguration.set(PATH, new ArrayList<>(set));
            save();
        }

        @Override
        public void save() {
            super.save();

            PlayerConfig.file = file;
            PlayerConfig.yamlConfiguration = this.yamlConfiguration;

            setAdvancementConfig(this);

            PlayerManager.reloadConfig(uuid);
        }
    }


    public void save() {
        try {
            yamlConfiguration.save(file);
            PlayerManager.reloadConfig(uuid);

            System.out.println(file.getName());
            System.out.println(hashCode());

            for (Map.Entry<UUID, PlayerConfig> entry : PlayerManager.getConfigs())
                System.out.println(entry.getKey() + ": " + entry.getValue().toString());

        } catch (IOException e) {
            Bukkit.getConsoleSender().sendMessage(MainAPI.getPrefix(MainAPI.PrefixType.ERROR) + "予期しないエラーが発生しました。\n" +
                    MainAPI.getPrefix(MainAPI.PrefixType.SECONDARY) + "位置: " + Main.PACKAGE_PATH + ".API.Manager.Player.PlayerConfig");
            e.printStackTrace();
        }
    }

    @Override
    public String toString() {
        return "UUID: " + getUUID().toString() + ", hasPermission: " + getPermission() + ", isBroadCaster: " + getBroadCaster();
    }
}

File: src/main/kotlin/jp/aoichaan0513/A_TosoGame_Live/API/Manager/Player/PlayerConfig.kt
package jp.aoichaan0513.A_TosoGame_Live.API.Manager.Player

import jp.aoichaan0513.A_TosoGame_Live.API.Enums.Advancement
import jp.aoichaan0513.A_TosoGame_Live.API.Enums.ItemType
import jp.aoichaan0513.A_TosoGame_Live.API.Interfaces.IConfig
import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI
import jp.aoichaan0513.A_TosoGame_Live.API.MainAPI.PrefixType
import jp.aoichaan0513.A_TosoGame_Live.API.Manager.World.WorldManager
import jp.aoichaan0513.A_TosoGame_Live.Main
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.configuration.file.YamlConfiguration
import java.io.File
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Paths
import java.util.*

class PlayerConfig(val uuid: UUID) {

    val file: File
    var config: YamlConfiguration

    var inventoryConfig: InventoryConfig
        private set
    var advancementConfig: AdvancementConfig
        private set

    init {
        val baseFileName = "player.yml"
        val fileName = "$uuid.yml"

        file = File("${Main.pluginInstance.getDataFolder()}${Main.FILE_SEPARATOR}players${Main.FILE_SEPARATOR}$fileName")
        if (file.exists()) {
            config = YamlConfiguration.loadConfiguration(file)
        } else {
            try {
                Files.copy(Main.pluginInstance.getResource(baseFileName), Paths.get("${Main.pluginInstance.getDataFolder()}${Main.FILE_SEPARATOR}players${Main.FILE_SEPARATOR}$fileName"))
            } catch (e: IOException) {
                e.printStackTrace()
            }
            config = YamlConfiguration.loadConfiguration(file)
        }

        inventoryConfig = InventoryConfig(file, config)
        advancementConfig = AdvancementConfig(file, config)
    }

    var difficulty: WorldManager.Difficulty
        get() = WorldManager.Difficulty.getDifficulty(config.getInt("difficulty", 1))
        set(difficulty) {
            config["difficulty"] = difficulty.id
            save()
        }

    var discordId: Long
        get() = config.getLong("discordId", 0)
        set(discordId) {
            config["discordId"] = discordId
            save()
        }

    var money: Long
        get() = config.getLong("money", 0)
        set(money) {
            config["money"] = money
            save()
        }

    var permission: Boolean
        get() = config.getBoolean("hasPermission", false)
        set(hasPermission) {
            config["hasPermission"] = hasPermission
            save()
        }

    var broadCaster: Boolean
        get() = config.getBoolean("isBroadCaster", false)
        set(isBroadCaster) {
            config["isBroadCaster"] = isBroadCaster
            save()
        }

    var bookForegroundColor: BookForegroundColor
        get() = BookForegroundColor.getColor(config.getString("bookForegroundColor") ?: "black")
        set(bookForegroundColor) {
            config["bookForegroundColor"] = bookForegroundColor.objectName
            save()
        }

    inner class InventoryConfig(f: File, c: YamlConfiguration) : IConfig(f, c) {
        private val PATH = "inventory"

        val items: Map<ItemType, Int>
            get() {
                val map = mutableMapOf<ItemType, Int>()

                for (key in config.getConfigurationSection(PATH)!!.getKeys(false)) {
                    val itemType = ItemType.getItemType(key)
                    map[itemType] = getSlot(itemType)
                }

                return map
            }

        private fun getSlot(key: String): Int {
            return getSlot(ItemType.getItemType(key))
        }

        fun getSlot(itemType: ItemType): Int {
            return c.getInt("$PATH.${itemType.itemName}", itemType.defaultSlot)
        }

        fun getSlot(slot: Int): ItemType? {
            val itemName = config.getConfigurationSection(PATH)!!.getKeys(false).firstOrNull { c.getInt("$PATH.${it}") == slot }
            return if (itemName != null) ItemType.getItemType(itemName) else null
        }

        fun setSlot(itemType: ItemType, slot: Int) {
            if (slot < 1) return

            val oldItem = getSlot(slot)
            if (oldItem != null)
                c["$PATH.${oldItem.itemName}"] = c.getInt("$PATH.${itemType.itemName}")
            c["$PATH.${itemType.itemName}"] = slot
            save()
        }

        override fun save() {
            super.save()
            inventoryConfig = this
        }
    }

    inner class AdvancementConfig(f: File, c: YamlConfiguration) : IConfig(f, c) {
        private val PATH = "advancements"

        fun addAdvancement(advancement: Advancement) {
            if (advancement == Advancement.UNKNOWN) return

            val set = c.getIntegerList(PATH).toMutableSet()
            set.add(advancement.id)
            c[PATH] = set.toList()
            save()
        }

        fun removeAdvancement(advancement: Advancement) {
            if (advancement == Advancement.UNKNOWN) return

            val set = c.getIntegerList(PATH).toMutableSet()
            set.remove(advancement.id)
            c[PATH] = set.toList()
            save()
        }

        fun hasAdvancement(advancement: Advancement): Boolean {
            return advancements.any { it.id == advancement.id }
        }

        var advancements: Set<Advancement>
            get() {
                val set = mutableSetOf<Advancement>()
                c.getIntegerList(PATH).filter { it != Advancement.UNKNOWN.id }.forEach { set.add(Advancement.getAdvancement(it)) }
                return set
            }
            set(advancements) {
                val set = mutableSetOf<Int>()
                advancements.filter { it != Advancement.UNKNOWN }.forEach { set.add(it.id) }
                c[PATH] = set.toList()
                save()
            }

        fun setAdvancements(vararg advancements: Advancement) {
            val set = mutableSetOf<Int>()
            advancements.filter { it != Advancement.UNKNOWN }.forEach { set.add(it.id) }
            c[PATH] = set.toList()
            save()
        }

        override fun save() {
            super.save()
            advancementConfig = this
        }
    }

    fun save() {
        try {
            config.save(file)
        } catch (e: IOException) {
            Bukkit.getConsoleSender().sendMessage("""
                ${MainAPI.getPrefix(PrefixType.ERROR)}予期しないエラーが発生しました。
                ${MainAPI.getPrefix(PrefixType.SECONDARY)}位置: ${Main.PACKAGE_PATH}.API.Manager.Player.PlayerConfig
            """.trimIndent())
            e.printStackTrace()
        }
    }

    enum class BookForegroundColor(val color: ChatColor) {
        BLACK(ChatColor.BLACK),
        WHITE(ChatColor.WHITE);

        val objectName: String
            get() = name.toLowerCase()

        companion object {

            fun getColor(name: String): BookForegroundColor {
                return values().firstOrNull { it.objectName.equals(name, true) } ?: BLACK
            }

            fun getColor(chatColor: ChatColor): BookForegroundColor {
                return values().firstOrNull { it.color == chatColor } ?: BLACK
            }
        }
    }
}